### PR TITLE
Add filter command

### PR DESCRIPTION
**Pull request by Austintjh19**

### CODE REVIEW

**Comment by Austintjh19** [comment]
Closed to rename package.

File: src/main/java/seedu/edulink/logic/parser/FilterCommandParser.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_TAG);
+
+        if (argMultimap.getAllValues(PREFIX_TAG).isEmpty()) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, FilterCommand.MESSAGE_USAGE));
+        }
+
+        tagList = ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG));
+
+        for (Tag tag : tagList) {
+            if (tag.tagName.trim().isEmpty()) {
```
Comment (dedsecrattle) [inline src/main/java/seedu/edulink/logic/parser/FilterCommandParser.java:39]: Can we pull out the logic of checking empty tags into a different method to increase readability

File: src/main/java/seedu/edulink/commons/util/StringUtil.java [hunk 2]
```diff
@@ -11,6 +11,30 @@
...
+    public static boolean matchesIgnoreCase(String sentence1, String sentence2) {
+        requireNonNull(sentence1);
+        requireNonNull(sentence2);
+
+        final String preppedSentence1 = sentence1.toLowerCase().trim();
+        final String preppedSentence2 = sentence2.toLowerCase().trim();
+
+        checkArgument(!preppedSentence2.isEmpty(), "Sentence parameter cannot be empty");
+
+        return preppedSentence1.equals(preppedSentence2);
```
Comment (Li-Zizhen) [inline src/main/java/seedu/edulink/commons/util/StringUtil.java:35]: I like how you extract logic into StringUtil.

File: src/test/data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json [hunk 3]
```diff
@@ -10,6 +10,8 @@
       "major": "Computer Science",
       "intake": "2023",
       "tags": [
+        "TA",
+        "Smart",
```
Comment (Li-Zizhen) [inline src/test/data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json:14]: Should we modify the typicalPersonAddressBook?

Comment (dedsecrattle) [reply src/test/data/JsonSerializableAddressBookTest/typicalPersonsAddressBook.json:14]: I think we shouldn't , this one is applicable only if the address book is empty

File: src/main/java/seedu/edulink/logic/commands/FilterCommand.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            + "the specified tags and displays them as a list with index numbers.\n"
+            + "Parameters: " + PREFIX_TAG + "KEYWORD "
+            + "[" + PREFIX_TAG + "KEYWORD] ... \n"
+            + "Example: " + COMMAND_WORD + " " + PREFIX_TAG + "TA" + " " + PREFIX_TAG + "Kind";
+
+    private final Predicate<Student> predicate;
+
+    public FilterCommand(Predicate<Student> predicate) {
+        this.predicate = predicate;
+    }
```
Comment (Li-Zizhen) [inline src/main/java/seedu/edulink/logic/commands/FilterCommand.java:31]: Should we put all the logic in FilerCommandParser.java? Would it be better to put some logic in FilterCommand.java?

File: src/test/java/seedu/edulink/logic/commands/FilterCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        Set<Tag> queryTags = new HashSet<>();
+        queryTags.add(testTag1);
+        queryTags.add(testTag2);
+
+        TagsContainQueryTagsPredicate predicate = new TagsContainQueryTagsPredicate(queryTags);
+        FilterCommand command = new FilterCommand(predicate);
+        expectedModel.updateFilteredPersonList(predicate);
+
+        assertCommandSuccess(command, model, expectedMessage, expectedModel);
+        assertEquals(Collections.emptyList(), model.getFilteredPersonList());
```
Comment (Li-Zizhen) [inline src/test/java/seedu/edulink/logic/commands/FilterCommandTest.java:86]: I like how you also take care of the filtered list instead of just checking models.

File: src/test/java/seedu/edulink/logic/parser/FilterCommandParserTest.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        assertParseSuccess(parser, "filter t/TA", expectedFilterCommand);
+
+        // Filter - 2 tags
+        Tag testTag2 = new Tag("Smart");
+        queryTags.add(testTag2);
+        assertParseSuccess(parser, "filter t/TA t/Smart", expectedFilterCommand);
+
+        // Filter - White Spaces
+        assertParseSuccess(parser, "filter t/ TA t/      Smart      ", expectedFilterCommand);
+    }
```
Comment (Li-Zizhen) [inline src/test/java/seedu/edulink/logic/parser/FilterCommandParserTest.java:56]: I like how you take care of the corner cases.

File: src/test/java/seedu/edulink/logic/commands/FilterCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        Set<Tag> queryTags = new HashSet<>();
+        queryTags.add(testTag1);
+
+        TagsContainQueryTagsPredicate predicate = new TagsContainQueryTagsPredicate(queryTags);
+        FilterCommand command = new FilterCommand(predicate);
+        expectedModel.updateFilteredPersonList(predicate);
+
+        assertCommandSuccess(command, model, expectedMessage, expectedModel);
+        assertEquals(Arrays.asList(ALICE, BENSON, DANIEL), model.getFilteredPersonList());
+    }
```
Comment (Li-Zizhen) [inline src/test/java/seedu/edulink/logic/commands/FilterCommandTest.java:121]: Should we also include some failed cases instead of only successful cases?

**Review by dedsecrattle — CHANGES_REQUESTED** [review]
Just Follow the Comments and make the necessary changes accordingly

File: src/main/java/seedu/edulink/model/student/TagsContainQueryTagsPredicate.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        boolean isMatch = true;
+        Set<Tag> studentTagList = student.getTags();
+
+        if (studentTagList.size() < filterTagList.size()) {
+            return false;
+        }
+
+        for (Tag filterTag: filterTagList) {
+            isMatch = studentTagList.stream()
+                    .anyMatch(studentTag -> StringUtil.matchesIgnoreCase(studentTag.tagName, filterTag.tagName));
```
[the remark this answers is not in the document]
Comment (Austintjh19) [reply src/main/java/seedu/edulink/model/student/TagsContainQueryTagsPredicate.java:32]: Thank you.

File: src/main/java/seedu/edulink/logic/commands/FilterCommand.java (hunk 4 above)
Comment (Austintjh19) [reply src/main/java/seedu/edulink/logic/commands/FilterCommand.java:31]: This is done to follow the implementation of the FindCommand which was an existent class beforehand. I think it is best to follow the same implementation as to not confuse future developers.

File: src/test/java/seedu/edulink/logic/commands/FilterCommandTest.java (hunk 7 above)
Comment (Austintjh19) [reply src/test/java/seedu/edulink/logic/commands/FilterCommandTest.java:121]: I believe it is covered in the test function name execute_nonExistentTags_noPersonFound().

**Review by dedsecrattle — APPROVED** [review]
